### ISSUE TITLE
iOS/Android Client

### ISSUE THREAD

**Issue by virl** [HelloZeroNet/ZeroNet#891]
How hard is it to make zeronet work as client on iOS/Android?

Is python support on these platforms would be enough?

**Comment by ysc3839** [comment]
There is an Android client. https://github.com/HelloZeroNet/ZeroNet-kivy
As for iOS, you can jailbreak and install python. The difficulty is that Apple's App Store does not allow apps runs dynamic code.

**Comment by mkg20001** [comment]
Actually with some changes the current android version _should_ run on iOS
We just need someone with a mac and iPhone who wants to write it

**Comment by virl** [comment]
@ysc3839 
> The difficulty is that Apple's App Store does not allow apps runs dynamic code

It allows to run dynamic code (for example, javascript & React), it just doesn't allow to update app's code by bypassing App Store updates.

**Comment by virl** [comment]
@ysc3839 
> As for iOS, you can jailbreak and install python

Can I (technically) run Python inside the app's process without jailbreaking?

**Comment by HelloZeroNet** [comment]
yes, eg.: https://itunes.apple.com/us/app/pythonista/id528579881?mt=8

**Comment by mishfit** [comment]
I think a complete rewrite in Swift/Java would be better.

A couple of things to consider:

1. System-wide TOR support allegedly coming soon to iOS, see [here](https://techstuffer.com/tor-for-ios-tor-is-coming-to-ios-soon/)
2. It should be possible to handle trackers, connections, etc natively
3. It would get you into the App Store or Google Play
4. Mobile clients could be leechers-only.
5. That would help standardize protocols/APIs and open up the possibility of having Java/ObjC/Node/C#/Ruby versions of ZeroNet clients.

If @shortcutme can add the repo to @HelloZeroNet I can start that (or I can create one and we can switch ownership afterwards).

**Comment by virl** [comment]
@mishfit Great analysis. I agree on all your points.

**Comment by grez911** [comment]
@mishfit Is there any progress?

**Comment by mkg20001** [comment]
@mishfit @grez911 I'm working on a nodejs/browser version (github.com/ZeroNetJS/ZeroNet-JS)

**Comment by grez911** [comment]
@mkg20001 Is it possible to run it on Android/IOS without any additional software?

**Comment by mkg20001** [comment]
Yes, but later. Currently some core features and the libp2p stuff need to be finished. Then it should be possible through libp2p-webrtc (similar to how it's being done [here](https://github.com/libp2p/js-libp2p/blob/master/examples/libp2p-in-the-browser/1/src/browser-bundle.js))
A demo of what works is already available at https://zeronetjs.github.io/znjs-web-bundle/ (but it only starts for now)

**Comment by tomachinz** [comment]
Does anybody know if it would be possible to: a) use the codebase for the Android client based on python still i understand..... in iOS but only personally, say without jailbrake but WITH side-loading via say XCode? I can load it onto each phone/tablet by hand, bypassing the app store? Or is this not possible (due to python I'm guessing)...

**Comment by mkg20001** [comment]
@tomachinz It seems like the android app [could be built for iOS too](https://github.com/kivy/buildozer#readme) but I have no idea how exactly it works and I couldn't even test it as it requires an iOS device

**Comment by BenMcLean** [comment]
ZeroNet really, really needs a working Android solution.

**Comment by imachug** [comment]
@BenMcLean Doesn't ZeroNet-Kiwy work?

**Comment by Serkan-devel** [comment]
Maybe publish the android Verizon on the F-Droid store already for the, ones who use custom roms or where Google Play is blocked.

**Comment by BenMcLean** [comment]
No because [the GUI's bugged](https://github.com/mkg20001/ZeroNet-kivy/issues/2).

**Comment by imachug** [comment]
It isn't bugged, it is not supported at all.

**Comment by mkg20001** [comment]
ZeroNet-kivy is in alpha, so don't expect this thing to work perfectly...

**Comment by emirzyxy** [comment]
Publishing it to App Store isn’t a problem, just use next.tweakboxapp.com or Cydia Impactor.

**Comment by rick-li** [comment]
https://github.com/rick-li/ZeroNetSwift  
I created a swift based poc, works for simple static site so far, 
i don't have time to maintain it, whoever want to continue the development please contact me I will transfer the ownership.

**Comment by Serkan-devel** [comment]
Maybe a react-native version. It can compile to iOS and Android

**Comment by NotS0n1cDev** [comment]
https://github.com/HelloZeroNet/ZeroNet/issues/891#issuecomment-366511279
Exactly what I was thinking. You keep saying it *needs* to be in the App Store, almost aggressively, but iOS allows for signing applications outside of the App Store, for free, or paid, depending.

The mentioned tool, Cydia Impactor, would work for free.
TweakBox (Signig Services in General) are paid for by the Signing Service, so it would be free for the end user.

(I can test by the way)

**Comment by SiNaPsEr0x** [comment]
or appdb for ipa...

**Comment by canewsin** [comment]
I'm currently working on more useful Android client with native code.

**Comment by canewsin** [comment]
App is Live on Playstore
Download Link : https://play.google.com/store/apps/details?id=in.canews.zeronet

**Comment by krzysztof113** [comment]
Good job canewsin.
It's not findable from search, did you publish fully?
@HelloZeroNet you can consider replacing kivy version with this.

**Comment by canewsin** [comment]
> Good job canewsin.
> It's not findable from search, did you publish fully?
> @HelloZeroNet you can consider replacing kivy version with this.

it needs some metadata, people need to search ZeroNet Mobile and select my app which is at end of list. after some recognisable downloads, it will be on top.

**Comment by krzysztof113** [comment]
Oh yes, it's last one

**Comment by HelloZeroNet** [comment]
Great, Thanks! Updated on the website: https://github.com/HelloZeroNet/zeronet.io/commit/94dce5ff2911a473f05bda5255fb98a268f15e15
